File: of06_从尾到头打印链表/Solution2.py
from typing import *

from util_py.list import *

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reversePrint(self, head: ListNode) -> List[int]:
        res = []

        def recurse(node):
            if not node:
                return
            recurse(node.next)
            res.append(node.val)
        recurse(head)
        return res


def test(test_name, head, expected):
    res = Solution().reversePrint(head)
    if res == expected:
        print(f'{test_name} success.')
    else:
        print(f'{test_name} failed.')


if __name__ == "__main__":
    head1 = build_list([1,3,2])
    expected1 = [2,3,1]
    test('test1', head1, expected1)



# 输入一个链表的头节点，从尾到头反过来返回每个节点的值（用数组返回）。

# 示例 1：

# 输入：head = [1,3,2]
# 输出：[2,3,1]

File: of06_从尾到头打印链表/Solution2.cpp
#include <iostream>
#include <vector>
#include "utils_cpp/list.h"
using namespace std;

class Solution {
public:
    vector<int> reversePrint(ListNode* head) {
        if (!head)
            return {};
        vector<int> res = reversePrint(head->next);
        res.push_back(head->val);
        return res;
    }
};


void test(string test_name, ListNode* head, vector<int> expected)
{
    vector<int> res = Solution().reversePrint(head);
    if (res == expected)
        cout << test_name << " success." << endl;
    else
        cout << test_name << " failed." << endl;
}

int main()
{
    ListNode* l1 = new ListNode(1);
    l1->next = new ListNode(3);
    l1->next->next = new ListNode(2);
    vector<int> expected1 = {2,3,1};
    test("test1", l1, expected1);

    return 0;
}
